src/tensorlake/function_executor/service.py

```diff
@@ -256,3 +357,159 @@ def get_info(
+            self._logger.error(
+                "task execution failed in background thread",
+                task_id=task_info.task.task_id,
+                exc_info=e,
+            )
+
+            raise
+
+        finally:
+            with task_info.execution.updated:
+                task_info.execution.complete = True
+                task_info.execution.updated.notify_all()
+
+    def list_tasks(
+        self, request: ListTasksRequest, context: grpc.ServicerContext
+    ) -> ListTasksResponse:
+        with self._tasks_lock:
+            tasks = []
+            for task_info in self._tasks.values():
+                with task_info.execution.updated:
+                    # Create a copy of the task without the request field for listing
+                    task_copy = Task()
+                    task_copy.CopyFrom(task_info.task)
+                    task_copy.ClearField(
+                        "request"
+                    )  # Don't return input data when listing
+                    task_copy.ClearField(
+                        "response"
+                    )  # Don't return output data when listing
+                    tasks.append(task_copy)
+
+            return ListTasksResponse(tasks=tasks)
+
+    def create_task(
+        self, request: CreateTaskRequest, context: grpc.ServicerContext
+    ) -> Task:
+        task = request.task
+
+        self._validate_task(task)
+
+        with self._tasks_lock:
+            if task.task_id in self._tasks:
+                context.abort(
+                    grpc.StatusCode.ALREADY_EXISTS,
+                    f"Task {task.task_id} already exists",
+                )
+
+            task_info = _TaskInfo(task)
+            self._tasks[task.task_id] = task_info
+
+            task_info.execution.thread = threading.Thread(
+                target=self._execute_task_in_thread,
+                args=(task_info,),
+                daemon=True,
+            )
+            task_info.execution.thread.start()
+
+        # Return a minimal task with no optional fields set; we'll
+        # extend this iff the server provides new info.
+        return Task()
+
+    def await_task(
+        self, request: AwaitTaskRequest, context: grpc.ServicerContext
+    ) -> Generator[AwaitTaskProgress, None, None]:
+        """Wait for task completion and stream progress updates."""
+        task_id = request.task_id
+
+        with self._tasks_lock:
+            task_info = self._tasks.get(task_id)
+
+        if task_info is None:
+            context.abort(
+                grpc.StatusCode.NOT_FOUND,
+                f"Task {task_id} not found",
+            )
+
+        # Stream progress updates until task completes
+        final_result = None
+        sent_first_progress = False
+        while True:
+            with task_info.execution.updated:
+                if task_info.execution.complete:
+                    # Use TaskResult directly
+                    final_result = task_info.task.result
```

Comment:
If task_info.task.result is None then Executor will treat this as internal error right?

Reply:
It had better - it can't depend on the function executor to behave correctly.

Reply:
I guess you mean yes, lgtm in this case.